tensorflow_lite_support/ios/task/audio/core/audio_record/sources/TFLAudioRecord.m

```diff
@@ -0,0 +1,269 @@
+                                                                   error:&conversionError])) {
+                      self->_globalError = conversionError;
+                    } else {
+                      self->_globalError = nil;
+                    }
+                  });
+                }];
+
+  [_audioEngine prepare];
+  [_audioEngine startAndReturnError:error];
+}
+
+- (BOOL)startRecordingWithError:(NSError **)error {
+  switch ([AVAudioSession sharedInstance].recordPermission) {
+    case AVAudioSessionRecordPermissionDenied: {
+      [TFLCommonUtils createCustomError:error
+                             withDomain:TFLAudioRecordErrorDomain
+                                   code:TFLAudioRecordErrorCodeRecordPermissionDeniedError
+                            description:@"Record permissions were denied by the user. "];
+      return NO;
+    }
+
+    case AVAudioSessionRecordPermissionGranted: {
+      [self startTappingMicrophoneWithError:error];
+      return YES;
+    }
+
+    case AVAudioSessionRecordPermissionUndetermined: {
+      [TFLCommonUtils
+          createCustomError:error
+                 withDomain:TFLAudioRecordErrorDomain
+                       code:TFLAudioRecordErrorCodeRecordPermissionUndeterminedError
+                description:@"Record permissions are undertermined. Yo must use AVAudioSession's "
+                            @"requestRecordPermission() to request audio record permission from "
+                            @"the user. Please read Apple's documentation for further details"
+                            @"If record permissions are granted, you can call this "
+                            @"method in the completion handler of requestRecordPermission()."];
+      return NO;
+    }
+  }
+}
+
+- (void)stop {
+  [[_audioEngine inputNode] removeTapOnBus:0];
+  [_audioEngine stop];
+  dispatch_sync(self->_conversionQueue, ^{
+    [self->_ringBuffer clear];
+  });
+}
+
+- (nullable TFLFloatBuffer *)readAtOffset:(NSUInteger)offset
+                                 withSize:(NSUInteger)size
+                                    error:(NSError *_Nullable *)error {
+  __block TFLFloatBuffer *bufferToReturn = nil;
+  __block NSError *readError = nil;
+
+  dispatch_sync(_conversionQueue, ^{
+    if (_globalError) {
+      readError = [_globalError copy];
+    } else if (offset + size > [_ringBuffer size]) {
+      [TFLCommonUtils
+          createCustomError:&readError
+                 withDomain:TFLAudioRecordErrorDomain
+                       code:TFLAudioRecordErrorCodeInvalidArgumentError
+                description:@"Index out of bounds: offset + size should be <= to the size of "
+                            @"TFLAudioRecord's internal buffer. "];
+    } else {
+      bufferToReturn = [_ringBuffer floatBufferWithOffset:offset size:size];
+    }
+  });
+
+  if (!bufferToReturn && error) {
+    *error = readError;
+  }
+  
+  return bufferToReturn;
+}
+
```

Comment:
Please write unit tests for this class. You can create a AVAudioPCMBuffer programmatically, feed it through the conversion pipeline and assert if `readAtOffset:withSize:error` returns a correct result. Please make sure to test both single-channel and double-channel input.

Reply:
Fixed.